lang/rust/avro/src/types.rs

```diff
@@ -1323,4 +1356,269 @@ mod tests {
+        outer_value.resolve(&schema).expect("Record schema defined in field must be resolvable in Record schema defined in other field");
+    }
+
+    #[test]
+    fn test_recursive_resolves5() {
+        let schema = Schema::parse_str(
+            r#"
+        {
+            "type":"record",
+            "name":"TestStruct",
+            "fields": [
+                {
+                    "name":"a",
+                    "type":{
+                        "type":"map",
+                        "values": {
+                            "type":"record",
+                            "name": "Inner",
+                            "fields": [ {
+                                "name":"z",
+                                "type":"int"
+                            }]
+                        }
+                    }
+                },
+                {
+                    "name":"b",
+                    "type": {
+                        "type":"array",
+                        "items":"Inner"
+                    }
+                }
+            ]
+        }"#,
+        )
+        .unwrap();
+
+        let inner_value1 = Value::Record(vec![("z".into(), Value::Int(3))]);
+        let inner_value2 = Value::Record(vec![("z".into(), Value::Int(6))]);
+        let outer_value = Value::Record(vec![
+            (
+                "a".into(),
+                Value::Map(vec![("akey".into(), inner_value2)].into_iter().collect()),
+            ),
+            ("b".into(), Value::Array(vec![inner_value1])),
+        ]);
+        outer_value
+            .resolve(&schema)
+            .expect("Record defined in map definition must be resolveable from array");
+    }
+
+    #[test]
+    fn test_recursive_resolves6() {
+        let schema = Schema::parse_str(
+            r#"
+        {
+            "type":"record",
+            "name":"TestStruct",
+            "fields": [
+                {
+                    "name":"a",
+                    "type":["null", {
+                        "type":"record",
+                        "name": "Inner",
+                        "fields": [ {
+                            "name":"z",
+                            "type":"int"
+                        }]
+                    }]
+                },
+                {
+                    "name":"b",
+                    "type":"Inner"
+                }
+            ]
+        }"#,
+        )
+        .unwrap();
+
+        let inner_value1 = Value::Record(vec![("z".into(), Value::Int(3))]);
+        let inner_value2 = Value::Record(vec![("z".into(), Value::Int(6))]);
+        let outer1 = Value::Record(vec![
+            ("a".into(), inner_value1),
+            ("b".into(), inner_value2.clone()),
+        ]);
+        outer1
+            .resolve(&schema)
+            .expect("Record definition defined in union must be resolvabled in other field");
+        let outer2 = Value::Record(vec![("a".into(), Value::Null), ("b".into(), inner_value2)]);
```

Comment:
The test fails because `a`s value is `Null` and there is no way to resolve the `Inner` type from it, so it is not known when `b` asks for it.

Reply:
Won't all named elements of the schema need to be visited (depth-first, left-to-right) by `resolve`, regardless of whether they're used in resolution, in order for `resolve` to work properly with references?

Reply:
Please check https://github.com/apache/avro/pull/1582/commits/09c2cb7bd1a75f65ef359751ff0937987d07b43d
It solves the current issue.

Reply:
@martin-g Thanks! That does seem to resolve the resolution issue for me. I'm now running into what I think is a similar problem during encoding while writing: it's crashing on [this `unwrap`](https://github.com/apache/avro/blob/44b8cd23756b752bf9d0976574453c712dcd44d9/lang/rust/avro/src/encode.rs#L62). I haven't looked closely, but it seems like exactly the same issue (if the value hasn't required the part of the schema with the definition to be looked at, the reference lookup fails).

In the longer term do you think it might be better to carry around the name-definition hash map in the schema, rather than having to rebuild it for each item in multiple places?

Reply:
I will improve it now!


> In the longer term do you think it might be better to carry around the name-definition hash map in the schema, rather than having to rebuild it for each item in multiple places?

Let's merge this PR once all the issues are fixed and then you or @jklamer can send a PR to re-work it the way you imagine. @jklamer also mentioned something similar at https://github.com/apache/avro/pull/1582#issuecomment-1060137762

Reply:
@travisbrown Should be fixed with https://github.com/apache/avro/pull/1582/commits/bd34804227c1da228595ff7b5a1f148944749ae9. Please try again!

Reply:
That seems to be working! Thanks again! I'll spend some more time with it tomorrow.

Reply:
Great!

@jklamer If you have time and desire please add some tests for encode.rs for the last fix, similar to the ones you added to types.rs.
It is late here. 
Thanks in advance!